src/sisl/io/_zipfile.py

```diff
@@ -0,0 +1,131 @@
+
+        self.close_zipfile = close_zipfile
+
+    @property
+    def suffix(self):
+        """Override the suffix property to ensure it returns a string"""
+        return Path(str(self)).suffix
+
+    def with_suffix(self, *args, **kwargs):
+        """Override the with_suffix method to ensure it returns a ZipPath"""
+
+        file_path = Path(str(self)).relative_to(self.root.filename)
+
+        return self.__class__(self.root, str(file_path.with_suffix(*args, **kwargs)))
+
+    def open(self, *args, **kwargs):
+        """Override the open method to patch the file handle if necessary"""
+
+        fh = super().open(*args, **kwargs)
+
+        if self.close_zipfile:
+            # The root zipfile needs to be closed when closing the file handle,
+            # so we wrap the close method of the file handle
+            close = fh.close
+
+            @functools.wraps(close)
+            def _close(*args, **kwargs):
+                close(*args, **kwargs)
+                self.root.close()
+
+            fh.close = _close
+
+        return fh
+
+    @classmethod
+    def from_zipfile_path(cls, zipfile_path: zipfile.Path, close_zipfile: bool = False):
+        """Create a ZipPath from a zipfile.Path object"""
+
+        file_path = Path(str(zipfile_path)).relative_to(zipfile_path.root.filename)
+
+        return cls(zipfile_path.root, str(file_path), close_zipfile=close_zipfile)
+
+    @classmethod
+    def from_path(cls, path: Path, mode: str = "r", close_zipfile: bool = True):
+        """Create a ZipPath from a Path object.
+
+        Given a path object, scans the path to find the first zip file in the path.
+        If a zip file is found, a ZipPath object is created.
+
+        This function initializes a new ``zipfile.ZipFile`` object to use as the root.
+
+        Parameters
+        ----------
+        path :
+            The path to the file or directory.
+        mode :
+            The mode in which you will want to open paths. This
+            determines the mode in which the zipfile is opened.
+        close_zipfile :
+            Whether to close the zipfile when closing the file handle.
+            This is important because if you write to a zipfile, you need
+            to close it to write the changes to disk.
+            Since the zipfile is created internally, the default is to close
+            it when we are done reading/writing to the path.
+        """
+        for i, part in enumerate(path.parts[:-1]):
+            if part.endswith(".zip"):
+                zip_path = Path(*path.parts[: i + 1])
+                if zip_path.is_file():
+                    zipfile_mode = "r" if mode.startswith("r") else "a"
+                    root_zip = zipfile.ZipFile(zip_path, zipfile_mode)
+                    return cls(
+                        root_zip,
+                        str(path.relative_to(zip_path)),
+                        close_zipfile=close_zipfile,
+                    )
```

Comment:
@zerothi we are already checking that the zip file exists before opening it, I didn't remember :sweat_smile: